[AArch64][GlobalISel] Combine G_EXTRACT_VECTOR_ELT and G_BUILD_VECTOR sequences into G_SHUFFLE_VECTOR

ValentijnvdBeek opened the pull request:
This combine tries to find all the build vectors whose source elements all originate from a G_EXTRACT_VECTOR_ELT from one or two donor vectors. One example where this may happen is for AI chips where there are a lot of matrix multiplications. Typically there vectors are dissected and then rearranged into the right transformation.

E.g.
```
  %donor1(<2 x s32>) = COPY $d0
  %donor2(<2 x s32>) = COPY $d1
  %ext1 = G_EXTRACT_VECTOR_ELT %donor1, 0
  %ext2 = G_EXTRACT_VECTOR_ELT %donor1, 1
  %ext3 = G_EXTRACT_VECTOR_ELT %donor2, 0
  %ext4 = G_EXTRACT_VECTOR_ELT %donor2, 1
  %vector = G_BUILD_VECTOR %ext1, %ext2, %ext3, %ext4
  ==>
  replace with:  %vector = G_SHUFFLE_VECTOR %donor1, %donor2, shufflemask(0, 1, 2, 3)
```

It has some speed ups in the AArch64 architecture. This pattern can also occur in other architectures were a lot of vector operations may happen. It can be tempting to express matrix operations in a similar fashion rather than as a one cohesive shufflevector directly.

ValentijnvdBeek committed:
[GlobalISel] Remove inaccurate input vector restriction

In the buildShuffleVector method, there is a restriction that the
input vectors must be larger than the mask. However, this is not
the definition of a ShuffleVector instruction that is used inside
of our test suite.

For example:
shuffle_concat_1 in combine_shuffle_vector.mir: 4xs8 -> 16xs8
v3s8_crash in legalize_insert_vector_elt: 3xs8 -> 12xs8
shuffle_vector_to_concat_vector_45670123 in prelegalizercombiner-shuffle-vector: 4xs32 -> 12xs32

ValentijnvdBeek committed:
[AArch64][GlobalISel] Combine G_EXTRACT_VECTOR_ELT and G_BUILD_VECTOR sequences into G_SHUFFLE_VECTOR

This combine tries to find all the build vectors whose source elements all originate from a G_EXTRACT_VECTOR_ELT from one or two donor vectors. One example where this may happen is for AI chips where there are a lot of matrix multiplications. Typically there vectors are disected and then  rearranged into the right transformation.
E.g.
  %donor1(<2 x s32>) = COPY $d0
  %donor2(<2 x s32>) = COPY $d1
  %ext1 = G_EXTRACT_VECTOR_ELT %donor1, 0
  %ext2 = G_EXTRACT_VECTOR_ELT %donor1, 1
  %ext3 = G_EXTRACT_VECTOR_ELT %donor2, 0
  %ext4 = G_EXTRACT_VECTOR_ELT %donor2, 1
  %vector = G_BUILD_VECTOR %ext1, %ext2, %ext3, %ext4
  ==>
  replace with:  %vector = G_SHUFFLE_VECTOR %donor1, %donor2, shufflemask(0, 1, 2, 3)

ValentijnvdBeek commented on llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:
```diff
@@ -4205,6 +4207,97 @@ void CombinerHelper::applyExtractVecEltBuildVec(MachineInstr &MI,
+  }
+
+  // We could reuse the same vector register and shuffle them both together
+  // but it is nicer for later optimizations to explicitely make it undef.
```
Nit: explicitely -> explicitly

ValentijnvdBeek commented on llvm/lib/CodeGen/GlobalISel/MachineIRBuilder.cpp:
```diff
@@ -772,8 +772,6 @@ MachineInstrBuilder MachineIRBuilder::buildShuffleVector(const DstOp &Res,
   LLT DstTy = Res.getLLTTy(*getMRI());
   LLT Src1Ty = Src1.getLLTTy(*getMRI());
   LLT Src2Ty = Src2.getLLTTy(*getMRI());
-  assert((size_t)(Src1Ty.getNumElements() + Src2Ty.getNumElements()) >=
```
As outlined in the commit, there are a couple of locations in the AArch64 where this assertion doesn't hold. Therefore, I removed it to bring the build function in line with the real(?) world usage of ShuffleVector.

https://github.com/llvm/llvm-project/blob/5d45815473496db4b041a008e60be17bd78c06ae/llvm/test/CodeGen/AArch64/GlobalISel/combine-shufflevector.mir#L35
https://github.com/llvm/llvm-project/blob/5d45815473496db4b041a008e60be17bd78c06ae/llvm/test/CodeGen/AArch64/GlobalISel/legalize-insert-vector-elt.mir#L285
https://github.com/llvm/llvm-project/blob/5d45815473496db4b041a008e60be17bd78c06ae/llvm/test/CodeGen/AArch64/GlobalISel/prelegalizercombiner-shuffle-vector.mir#L192

tschuett replied:
Why did this assert start failing with your PR? It seemed to work before.

tschuett replied:
I/we build our LLVM for development with `-DLLVM_ENABLE_ASSERTIONS=On`. If an assert hits, my first thought is: Oh wait, I broke something. My thought is not: this is a legacy assert and I should remove it. Please don't remove asserts. You should think about fixing your code/contribution.

ValentijnvdBeek replied:
You are completely right, that the code should be checked with asserts and that they are an important part of ensuring quality going further. Thanks for the reminder, I think that is never unwelcome. However, in this case my PR doesn't break the assertion per se. Rather the examples of `G_SHUFFLE_VECTOR`, I mentioned before don't follow the assertion. The code asserts that the sum of the vectors is larger than the mask, but for none of those are true. Since those patterns exist, I want to support them and since we have a `buildShuffleVector` I want to use it (instead of reimplementing it). 

Anyways, I am completely willing to drop the commit, you know much more than me about things like this. But my reason for removing it was to pull into line with how the opcode is used, not because I thought it was a legacy assert :)

arsenm replied:
The source vectors each must be the same size, and the result number of elements must exactly match the number of elements in the mask. The assert doesn't quite match the machine verifier's rules for G_SHUFFLE_VECTOR

ValentijnvdBeek commented on llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:
```diff
@@ -4205,6 +4207,97 @@ void CombinerHelper::applyExtractVecEltBuildVec(MachineInstr &MI,
+  VectorRegisters =
+      std::make_pair(RegisterVector.front(), RegisterVector.back());
+
+  // We check that they're the same type before running. We can also grow the
+  // smaller one to the target size, but there isn't an elegant way to do that
+  // until we have a good lowering for G_EXTRACT_SUBVECTOR.
+  if (MRI.getType(VectorRegisters.first) != MRI.getType(VectorRegisters.second))
+    return false;
+
```
In the best case scenario, you would cast up the lower one into the right size by adding `G_CONCAT_VECTOR` instructions until they are in the same size. A few months ago the `G_EXTRACT_SUBVECTOR` was added and, very recently, a lot of work has gone into implementing it for the RISCV backend. But I don't think there is a nice fallback legalization strategy, which is what I would want before relying on it in the Combiner.

What has worked well (?) for me before is adding `G_CONCAT_VECTOR` and `G_UNMERGE` until the target has been reached. But, I am not sure if that is the best strategy. If anybody has a good idea, please let me know so that I can work on it.

tschuett commented on llvm/include/llvm/Target/GlobalISel/Combine.td:
```diff
@@ -905,6 +905,16 @@ def extract_vec_elt_build_vec : GICombineRule<
+  GIDefMatchData<"std::pair<Register, Register>">;
+def extract_vector_element_build_vector_to_shuffle_vector : GICombineRule<
+  (defs root:$root, extract_vector_register_to_id_mapping_matchinfo:$matchinfo, vector_reg_pair_matchinfo:$regpair),
+  (match (wip_match_opcode G_BUILD_VECTOR):$root,
```
There will be comments about style and algorithms. What I really don't like is the pattern. The pattern matches on build vectors, but the combine is about build vectors of extract vector elts.

```
def extract_vector_element_build_to_shuffle_vector2 : GICombineRule<
   (defs root:$root, build_fn_matchinfo:$matchinfo),
   (match
          (G_EXTRACT_VECTOR_ELT $el2, $vec2, $idx2),
          (G_EXTRACT_VECTOR_ELT $el1, $vec1, $idx1),
          (G_BUILD_VECTOR $root, $el1, $el2):$build,
   [{ return Helper.matchCombineExtractToShuffle(*${build}, ${matchinfo}, ${regpair}); }]),
   ( apply [{ Helper.applyCombineExtractToShuffle(*${build}, ${matchinfo}, ${regpair}); }] )>;
```
This a matcher, but we would copy and paste for many build vector sizes ...

tschuett replied:
https://llvm.org/docs/GlobalISel/MIRPatterns.html#gallery

ValentijnvdBeek replied:
Thanks for the review, sorry for committing something that you really don't like. I hope that are things about it you do enjoy. I really appreciate the link. I missed it when I looked for something like it before and I am happy with any information about tablegen that I can find.

In this case, I am not entirely sure if it is exactly the right solution. As you mention, there are a lot of sizes to take into account. For the backend that I am the most for familiar with, AI Engine, the vector size is a maximum of 2048 and a minimum scalar size of 8. So that has 256 patterns to match for both this combiner and the three around it (that use the same pattern). I think we only need to do the powers of 2 and 3 (AMDGPU), so that makes it about ~14 patterns or ~42 in total. 

Another reservation that I have is that in this case, complete enumeration is a bit difficult. There is always possible that there is an architecture that has a vector size that just a bit larger or has a different size than the others. By enumerating explicitly, future/alternatives backends need to remember/know to update this combiner specifically. 

However, this is my first contribution to LLVM. So these are just, somewhat layman, considerations. If you think that they are not as relevant as I think, than I more willing to change the pattern and also update the combiners around it with the same pattern.

tschuett replied:
We are interested in compile-time. The combiner engine is more efficient than the C++ code. That is why we strive for precise patterns.  If we jump for every build vector into the C++ code, the success rate of the C++ code will be extremely low. It is waste of compile-time. For precise patterns, we only jump into C++ code when there is good chance to succeed. If the pattern of this combine is build vector and the chance of success of the combine is really low, then we have to try something else.

aemerson replied:
> Thanks for the review, sorry for committing something that you really don't like. I hope that are things about it you do enjoy. I really appreciate the link. I missed it when I looked for something like it before and I am happy with any information about tablegen that I can find.

Please don't be discouraged, Thorsten was only referring to the approach itself and it's common and expected in LLVM code reviews to critique the code in that manner. It's not a reflection on whether we want your contribution at all, usually it's just a matter of making sure it's done in the best way possible. We'd love to have more people like yourself contribute to the project!

I haven't thought about this patch in detail yet but once I do I'll give my feedback.

ValentijnvdBeek replied:
@tschuett I fixed this in a fixup commit for powers of two (from what I could find also AMDGPU uses those as registers sizes). There seems to be a cap for the amount of patterns that can be supported (1024x8, 2048x16, 4096x32) and it does add a lot of extra lines. Just this pattern now takes 25-30% of the file and patterns around it should also be changed to this.

arsenm commented on llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:
```diff
@@ -4205,6 +4207,97 @@ void CombinerHelper::applyExtractVecEltBuildVec(MachineInstr &MI,
+bool CombinerHelper::matchCombineExtractToShuffle(
+    MachineInstr &MI, SmallVectorImpl<std::pair<Register, int>> &VecIndexPair,
+    std::pair<Register, Register> &VectorRegisters) {
+  assert(MI.getOpcode() == TargetOpcode::G_BUILD_VECTOR);
```
This is redundant with the cast

ValentijnvdBeek replied:
Replaced in a fixup commit (to be squashed before merging)

arsenm commented on llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:
```diff
@@ -4205,6 +4207,97 @@ void CombinerHelper::applyExtractVecEltBuildVec(MachineInstr &MI,
+  const GBuildVector *BuildVector = cast<GBuildVector>(&MI);
+  Register SecondRegister = VectorRegisters.second;
+  if (FirstRegister == SecondRegister) {
+    SecondRegister = MRI.createGenericVirtualRegister(FirstRegisterType);
+    Builder.buildUndef(SecondRegister);
```
Avoid directly using MRI.createGenericVirtualRegister, can pass the type to createUndef to create the register

ValentijnvdBeek replied:
I didn't know that, thanks for pointing it out! I replaced it.

arsenm commented on llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:
```diff
@@ -4205,6 +4207,97 @@ void CombinerHelper::applyExtractVecEltBuildVec(MachineInstr &MI,
+  // but it is nicer for later optimizations to explicitely make it undef.
+  const GBuildVector *BuildVector = cast<GBuildVector>(&MI);
+  Register SecondRegister = VectorRegisters.second;
+  if (FirstRegister == SecondRegister) {
```
Probably should also have a separate combine for this

ValentijnvdBeek replied:
Thanks for the review! What exactly should be a separate combine? This sequence with one source? Or turning shufflevectors with the same source registers into a source and an undefined register?

arsenm replied:
Shufflevectors with the same source registers into a source and undefined (this came up on another review recently too)

ValentijnvdBeek replied:
Sure, I will remove the check from the current PR. Sounds like a great target for me to work on when I have a bit of free time, thanks for the suggestion! I appreciate it.

arsenm commented on llvm/test/CodeGen/AArch64/GlobalISel/combine-build-vector.mir:
```diff
@@ -151,3 +151,305 @@ body:             |
+    %11:_(s32) = G_EXTRACT_VECTOR_ELT %1:_(<4 x s32>), %2:_(s64)
+    %12:_(<2 x s32>) = G_BUILD_VECTOR %10:_(s32), %11:_
+    RET_ReallyLR implicit %12
+...
```
Test with an unexpected other user

ValentijnvdBeek replied:
I think that I did what you asked. I added a commit where one of the arguments is not a `G_EXTRACT_VECTOR_ELT` and on where it is a `G_BUILD_VECTOR_TRUNC` (which I don't target yet). But I am not entirely sure what you mean with other user. I must I admit that I am not entirely familiar with the LLVM lingo yet :)

ValentijnvdBeek replied:
Is this resolved with the new commit, then I can close it.

ValentijnvdBeek committed:
Fixup! Remove redudant G_BUILD_VECTOR assert

ValentijnvdBeek committed:
Fixup! Add aditional tests

ValentijnvdBeek committed:
Fixup! Create undef using buildUndef

tschuett commented:
There are no legality checks.

aemerson commented:
So while I see some nice code quality improvements in the tests, I'm not convinced this is a good transformation to make in the general case. The problem is that vector extracts->inserts are simple operations and a shuffle is in general a large and expensive operation, unless it can be pattern matched to a more precise variant like zip. For AArch64, `tbl` is an expensive operation that requires a load from a constant pool to materialize the indices, as well as the actual `tbl` itself.

In your example in this PR description, your input could have been optimized instead into a G_CONCAT_VECTOR right? More specific operations are usually faster and they're also easier for other optimizations to reason about. Shuffles suffer from the problem that they're more opaque unless we do expensive analysis.

ValentijnvdBeek commented on llvm/test/CodeGen/AArch64/arm64-neon-copy.ll:
```diff
@@ -1923,49 +1903,12 @@ entry:
-; CHECK-GI-NEXT:    mov v2.b[15], v0.b[0]
-; CHECK-GI-NEXT:    mov v0.16b, v2.16b
-; CHECK-GI-NEXT:    ret
+; CHECK-LABEL: test_concat_v16i8_v8i8_v8i8:
+; CHECK:       // %bb.0: // %entry
+; CHECK-NEXT:    // kill: def $d0 killed $d0 def $q0
+; CHECK-NEXT:    // kill: def $d1 killed $d1 def $q1
+; CHECK-NEXT:    mov v0.d[1], v1.d[0]
+; CHECK-NEXT:    ret
```
This turns into the following Generic IR, by first turning it into a shufflevector and then running the G_CONCAT_VECTORS [shufflevector combiner](https://github.com/llvm/llvm-project/blob/6c7a3f80e75de36f2642110a077664e948d9e7e3/llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp#L475). 

```llvm
bb.1.entry:
  liveins: $d0, $d1
  %0:_(<8 x s8>) = COPY $d0
  %1:_(<8 x s8>) = COPY $d1
  %51:_(<16 x s8>) = G_CONCAT_VECTORS %0:_(<8 x s8>), %1:_(<8 x s8>)
  $q0 = COPY %51:_(<16 x s8>)
  RET_ReallyLR implicit $q0
```

ValentijnvdBeek commented:
First of thanks for the great detailed comment Amara, I don't have a lot of experience with AArch64, so your perspective is very helpful.

> So while I see some nice code quality improvements in the tests, I'm not convinced this is a good transformation to make in the general case. The problem is that vector extracts->inserts are simple operations and a shuffle is in general a large and expensive operation, unless it can be pattern matched to a more precise variant like zip.

The combiner based on the lowering of the shuffle vector. When you lower a shuffle vector it will [turn into a sequence of extract->buildvector](https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/GlobalISel/LegalizeHelper.cpp#L8065) operations which is exactly what I match here. So any backend either uses this lowering and performs the same, or has a more efficient implementation of shuffle vector and performs better. However, I am not sure if this assumption actually holds in reality, so hopefully you can shine a light on that. 

> In your example in this PR description, your input could have been optimized instead into a G_CONCAT_VECTOR right?

You are right and it does actually. [Another combiner](https://github.com/llvm/llvm-project/blob/6c7a3f80e75de36f2642110a077664e948d9e7e3/llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp#L475) turns the shufflevector into a `G_CONCAT_VECTORS`, which is a consequence of this combiner being there. This is where the performance improvements in the tests come from. For example, it turns the large sequence of operations in [arm64-neon-copy.ll](https://github.com/llvm/llvm-project/pull/110545/commits/d3edbb416eab0ada38ac0146f1de61507039e608#diff-7f02a23a0657bd646f3015c3e828c7cd50f0cda5aec72f4536766f97b4c67cf2L1925) into a shufflevector which is then turned into a `G_CONCAT_VECTORS`

I didn't mention it since it a consequence of another combiner kicking in, although this is an intentional effect. Probably should have put a comment explaining the speedups for the cases where it conforms to the Selection DAG output, but I didn't want to overload the PR with additional comments. Apologies for the confusion I caused by that oversight.

> More specific operations are usually faster and they're also easier for other optimizations to reason about. Shuffles suffer from the problem that they're more opaque unless we do expensive analysis.

The background of the PR is that I worked during my internship on extending the analysis of shufflevectors and replacing them with more specific opcodes. By pulling these sequences into a shufflevector, the analysis is run on it and then either the sequence is replaced by an equivalent opcode or the shufflevector is lowered, in the generic case, into the exact same sequence as it was before. You mention that is the analysis is expensive, which is true, but we are already running this analysis for [shufflevectors](https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp#L475) and I am hoping that the speed improvements in the generated code will make up for the additional compile overhead.

With some luck my previous company will let me upstream that code and allow me to implement some of the ideas I have with regards of reducing the cost of this analysis. But no promises.

aemerson commented:
> First of thanks for the great detailed comment Amara, I don't have a lot of experience with AArch64, so your perspective is very helpful.
> 
> > So while I see some nice code quality improvements in the tests, I'm not convinced this is a good transformation to make in the general case. The problem is that vector extracts->inserts are simple operations and a shuffle is in general a large and expensive operation, unless it can be pattern matched to a more precise variant like zip.
> 
> The combiner based on the lowering of the shuffle vector. When you lower a shuffle vector it will [turn into a sequence of extract->buildvector](https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/GlobalISel/LegalizeHelper.cpp#L8065) operations which is exactly what I match here. So any backend either uses this lowering and performs the same, or has a more efficient implementation of shuffle vector and performs better. However, I am not sure if this assumption actually holds in reality, so hopefully you can shine a light on that.

Yes that's true, but let's think about this from first principles. In LLVM IR, we don't have separate instructions for things like vector concat, zip, etc... because the general shuffle can implement those. We rely on optimizations later on in codegen to select the most optimal instructions for some patterns.

In the GISel MIR, we don't have the same philosophy as the IR. Our goal is not to have a minimal MIR but to instead aid in lowering and selecting good target instructions. So we decompose some complex operations that targets don't have native support for into simpler ops, because now one of the factors we have to take into account is the idea of instruction legality.

If we were to perform this transform, we would be going from simpler (but more numerous) operations into more complex single operation. In order to get performant output, we rely on another transform to then re-optimize our output into the idea form. Therefore, our transform is not so much an optimization but instead a form of canonicalization.

If we decided to instead directly pattern match into a vector concat for example, we're skipping the canonicalization step and going straight to the instruction we want to emit. There's no ambiguity here about what's the best final output, in this case it's always a G_CONCAT_VECTOR. So going directly has benefits in 1) not having to spend time going through the intermediate step, and b) having the guaranteed output generated.

> 
> > In your example in this PR description, your input could have been optimized instead into a G_CONCAT_VECTOR right?
> 
> You are right and it does actually. [Another combiner](https://github.com/llvm/llvm-project/blob/6c7a3f80e75de36f2642110a077664e948d9e7e3/llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp#L475) turns the shufflevector into a `G_CONCAT_VECTORS`, which is a consequence of this combiner being there. This is where the performance improvements in the tests come from. For example, it turns the large sequence of operations in [arm64-neon-copy.ll](https://github.com/llvm/llvm-project/pull/110545/commits/d3edbb416eab0ada38ac0146f1de61507039e608#diff-7f02a23a0657bd646f3015c3e828c7cd50f0cda5aec72f4536766f97b4c67cf2L1925) into a shufflevector which is then turned into a `G_CONCAT_VECTORS`
> 
> I didn't mention it since it a consequence of another combiner kicking in, although this is an intentional effect. Probably should have put a comment explaining the speedups for the cases where it conforms to the Selection DAG output, but I didn't want to overload the PR with additional comments. Apologies for the confusion I caused by that oversight.
> 
> > More specific operations are usually faster and they're also easier for other optimizations to reason about. Shuffles suffer from the problem that they're more opaque unless we do expensive analysis.
> 
> The background of the PR is that I worked during my internship on extending the analysis of shufflevectors and replacing them with more specific opcodes. By pulling these sequences into a shufflevector, the analysis is run on it and then either the sequence is replaced by an equivalent opcode or the shufflevector is lowered, in the generic case, into the exact same sequence as it was before. You mention that is the analysis is expensive, which is true, but we are already running this analysis for [shufflevectors](https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp#L475) and I am hoping that the speed improvements in the generated code will make up for the additional compile overhead.
> 
> With some luck my previous company will let me upstream that code and allow me to implement some of the ideas I have with regards of reducing the cost of this analysis. But no promises.
Improving pattern matching of shuffle idioms so that we lower into more optimal instructions sounds like a good idea. I don't think that means that we need to canonicalize everything into shufflevectors however.

ValentijnvdBeek committed:
[GISel] Factor out the mask matching code from the shufflevector combiner

arsenm commented:
FWIW the DAG does try to do this: https://github.com/llvm/llvm-project/blob/f0ed31ce4b63a5530fd1de875c0d1467d4d2c6ea/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L24096

I can see this pattern appearing if some operations require some degree of vector splitting, and others do not

aemerson commented:
> FWIW the DAG does try to do this: https://github.com/llvm/llvm-project/blob/f0ed31ce4b63a5530fd1de875c0d1467d4d2c6ea/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L24096
> 
> I can see this pattern appearing if some operations require some degree of vector splitting, and others do not

Yeah but I don't think it's a good transformation to do if we can express the shuffle in a simpler & equally compact way. I can see that being helpful for long, complex sequences of extracts/inserts but if we know the idiom (e.g. concat) we should go directly.

ValentijnvdBeek committed:
fixup! Directly run the shuffle vector analysis code

ValentijnvdBeek commented on llvm/include/llvm/Target/GlobalISel/Combine.td:
```diff
@@ -926,8 +926,8 @@ def extract_all_elts_from_build_vector : GICombineRule<
-  extract_all_elts_from_build_vector,
-  extract_vector_element_build_vector_to_shuffle_vector]>;
+  extract_vector_element_build_vector_to_shuffle_vector,
+  extract_all_elts_from_build_vector]>;
```
nit: useless change, will be removed when squashed so this can be ignored

ValentijnvdBeek commented on llvm/test/CodeGen/AArch64/arm64-rev.ll:
```diff
@@ -397,13 +397,10 @@ define void @test_vrev64(ptr nocapture %source, ptr nocapture %dst) nounwind ssp
-; CHECK-GI-NEXT:    str h1, [x1, #2]
+; CHECK-GI-NEXT:    add x8, x1, #2
+; CHECK-GI-NEXT:    st1.h { v0 }[6], [x1]
+; CHECK-GI-NEXT:    st1.h { v0 }[5], [x8]
```
This is inline with @aemerson comment about `tbl`, it no longer turns non-optimizable sequences into shufflevectors so it keeps previous iteration of the code.

ValentijnvdBeek commented:
> If we decided to instead directly pattern match into a vector concat for example, we're skipping the canonicalization step and going straight to the instruction we want to emit. There's no ambiguity here about what's the best final output, in this case it's always a G_CONCAT_VECTOR. So going directly has benefits in 1) not having to spend time going through the intermediate step, and b) having the guaranteed output generated.

Thanks for your clear explanation of GISelm MIR, I appreciate the time you took and I understand a lot better now. I have updated the code so that it directly runs the analysis rather than doing a step in the middle. That has the benefit that we can still share code without actually having to turn into shufflevector.

> FWIW the DAG does try to do this: https://github.com/llvm/llvm-project/blob/f0ed31ce4b63a5530fd1de875c0d1467d4d2c6ea/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L24096
>
> I can see this pattern appearing if some operations require some degree of vector splitting, and others do not

>Yeah but I don't think it's a good transformation to do if we can express the shuffle in a simpler & equally compact way. I can see that being helpful for long, complex sequences of extracts/inserts but if we know the idiom (e.g. concat) we should go directly.

I mean it does both. The PR aims to simplify long, complex sequences that occur in matrix operations like the ones that were common for the backend that I used to work on (as mentioned in the PR description). With the side benefit of the code improvements that the shufflevector optimizations bring. Since this is not the backend I worked on and I don't how common it is in other backends, it isn't my focus atm. 

But, both of you know more than I ever forget, so I will eagerly await the result of this discussion. One suggestion that I have is that, in line with the comment from @tschuett, I can use specific combiners for each size. Smaller sizes are optimized directly, but larger sequences (say >32) are turned into shufflevector. Does that sound like an idea?

arsenm commented on llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:
```diff
@@ -4205,6 +4218,76 @@ void CombinerHelper::applyExtractVecEltBuildVec(MachineInstr &MI,
   replaceSingleDefInstWithReg(MI, Reg);
 }
 
+bool CombinerHelper::matchCombineExtractToShuffle(
```
I suggest to follow along with the DAG version; you'll want something similar to buildLegalVectorShuffle, which will avoid creating shuffles which will just be broken up again https://github.com/llvm/llvm-project/blob/33c14f19656c751bbbc083e4a168ab898e583bfd/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L23501

ValentijnvdBeek replied:
Thanks for the suggestion, I am now avoiding making a shufflevector after input from Amara. But I will be sure to check this if I need to make a shufflevector in the future.

aemerson commented on llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:
```diff
@@ -472,39 +474,16 @@ bool CombinerHelper::tryCombineShuffleVector(MachineInstr &MI) {
-  //       we could still emit an extract vector element in that case.
-  if (DstNumElts < 2 * SrcNumElts && DstNumElts != 1)
-    return false;
+bool CombinerHelper::matchVectorMaskSequence(
```
Since this is now a shared utility, could you give it a more specific name and also add a doc comment to explain what it does.

ValentijnvdBeek replied:
Added a comment and changed the name slightly, it is still a bit broad to make sure that it doesn't get pidgeonholed as just concatenation. Please tell me if you would like it to be changed

aemerson commented on llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:
```diff
@@ -4205,6 +4218,76 @@ void CombinerHelper::applyExtractVecEltBuildVec(MachineInstr &MI,
+
+  // This combine tries to find all the build vectors whose source elements
+  // all originate from a G_EXTRACT_VECTOR_ELT from one or two donor vectors.
+  // One example where this may happen is for AI chips where there are a lot
+  // of matrix multiplications. Typically there vectors are disected and then
```
We don't particularly attach specific use cases like "AI chips" to comments, the transform description should be enough.

ValentijnvdBeek replied:
Removed

aemerson reviewed (approved):
LGTM with some nits, thanks!

ValentijnvdBeek committed:
Nit! remove use case from comment

ValentijnvdBeek committed:
Fixup! Add comment to function

ValentijnvdBeek committed:
Fixup! Use the gallery MIR pattern to be more precise

ValentijnvdBeek committed:
Fixup! Rename function to a clearer name

ValentijnvdBeek commented:
> LGTM with some nits, thanks!

I addresses the nits. Thanks for the great feedback that you gave me, they were really helpful and I learned about LLVM from them!
